wgpu-hal/src/dx12/mod.rs

```diff
@@ -147,104 +175,102 @@ impl D3D12Lib {
                 "Root signature serialization error: {:?}",
                 unsafe { error.as_c_str() }.unwrap().to_str().unwrap()
             );
-            return Err(crate::DeviceError::Lost);
+            return Err(crate::DeviceError::Unexpected); // could be hal_usage_error or hal_internal_error
         }
 
-        Ok(D3DBlob(blob.expect(
-            "D3D12SerializeRootSignature succeeded but result is NULL?",
-        )))
+        blob.ok_or(crate::DeviceError::Unexpected)
     }
 
-    fn debug_interface(
-        &self,
-    ) -> Result<windows::core::Result<Direct3D12::ID3D12Debug>, libloading::Error> {
+    fn debug_interface(&self) -> Result<Direct3D12::ID3D12Debug, crate::DeviceError> {
         // Calls windows::Win32::Graphics::Direct3D12::D3D12GetDebugInterface on d3d12.dll
         type Fun = extern "system" fn(
             riid: *const windows_core::GUID,
             ppvdebug: *mut *mut core::ffi::c_void,
         ) -> windows_core::HRESULT;
-        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"D3D12GetDebugInterface") }?;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"D3D12GetDebugInterface\0") }?;
+
+        let mut result__ = None;
+
+        (func)(&Direct3D12::ID3D12Debug::IID, <*mut _>::cast(&mut result__))
+            .ok()
+            .into_device_result("GetDebugInterface")?;
 
-        let mut result__ = core::ptr::null_mut();
-        Ok((func)(&Direct3D12::ID3D12Debug::IID, &mut result__)
-            .and_then(|| unsafe { windows_core::Type::from_abi(result__) }))
+        result__.ok_or(crate::DeviceError::Unexpected)
```

Comment:
Note that a lot of this was intentionally verbatim-copied from `windows-rs` to allow easier updating.

Reply:
I see, but I think it's more valuable to have those be consistent. I'm also not a fan of using the `from_abi` function which maps null to `Error::empty()`.